### PR TITLE
Add Woo Subscription Stripe Billing settings and migration status notices

### LINKED ISSUE #6508: Enable merchants to opt into Stripe Billing as subscription engine
### Description

As part of our changes to the WooPayments plugin, we will be removing the subscription features that were previously bundled within it, referred to as WC Pay Subscriptions. However, we want to assure you that we will still offer Stripe Billing as an optional billing engine for merchants who choose to use it on their WC Pay stores.

To enable merchants to switch between these billing engines, we are introducing a new setting that allows the merchant to opt into Stripe Billing as the engine.

### Acceptance criteria

* A setting displayed on **WooCommerce > Settings > Payment** screen when WC Subscriptions plugin is active.
   * Disabled by default.
   * Positioned inline with other settings instead of being hidden under Advanced settings until there is a separate promotional flow for Stripe Billing.
* Includes an explanation of Stripe Billing including:
   * Off-site billing engine: in the most merchant friendly way possible explain how Stripe Billing offers a way to have payments processed on Stripe’s infrastructure instead of the merchant’s own WooCommerce site, with a brief pros/cons of that.
   * Limitations: only supports card payments, no https://github.com/Automattic/woocommerce-payments/issues/5677 etc.
   * Link to detailed documentation to learn more, like [this existing doc](https://href.li/?https://woocommerce.com/document/woocommerce-payments/built-in-subscriptions/comparison/#billing-engine).

### Designs

To come. 

### Additional context

peNR48-2u-p2 project.

### PR DESCRIPTION
**Pull request by james-allan**
Fixes #6508

### Changes proposed in this Pull Request

This PR adds a new "Subscriptions" setting section to the WooPayments setting screen. This new section is only shown if you have the Woo Subscriptions plugin active. 

It contains a simple checkbox to enable/disable Stripe Billing along with an array of notices that are shown depending on the Stripe Billing migration status. 

<p align="center">
Screenshot 2023-09-06 at 1 20 50 pm
</p>

### Testing

<details><summary><strong>No Settings</strong></summary>
<p>

**Setup:**

1. No Woo Subscriptions plugin installed. 

**Testing instructions:**

1. Go to **WooCommerce → Settings → Payments → Woo Payments**
2. Note there isn't any **"Subscriptions"** and Stripe Billing setting shown. 

</p>
</details>

<details><summary><strong>No Stripe Billing subscriptions</strong></summary>
<p>

**Setup:**

1. Woo Subscriptions plugin installed. 
2. No pre-existing Stripe Billing subscriptions. 

**Testing instructions:**

1. Go to **WooCommerce → Settings → Payments → Woo Payments**
2. Note there is a new **"Subscriptions"** section in the **Advanced** section with a single setting.
   - The setting should be disabled by default. 
3. Enable/disabling the setting and saving should set the `_wcpay_feature_stripe_billing` option to `1`/`0`.

Screenshot 2023-09-06 at 2 48 13 pm

</p>
</details>

<details><summary><strong>Enable Stripe Billing</strong></summary>
<p>

**Setup:**

1. Woo Subscriptions plugin installed. 
2. A subscription product of any type. 

**Testing instructions:**

1. Go to **WooCommerce → Settings → Payments → Woo Payments**
3. Enable the Stripe Billing setting.
4. Purchase a subscription. 
5. Check that the subscription is set up with Stripe Billing.
    - Go to **WooCommerce → Subscriptions**
    - Edit the subscription. 
    - You should notice it has a WCPay subscription ID.
6. This subscription should behave like a Stripe Billing subscription.

Screenshot 2023-08-31 at 3 06 00 pm

</p>
</details>

<details><summary><strong>Disable Stripe Billing with Subscriptions</strong></summary>
<p>

**Setup:**

1. Woo Subscriptions plugin installed. 
2. Stripe Billing enabled.
3. At least 1 Stripe Billing subscription from:
   - A previous test,
   - Created with WCPay Subscriptions built in functionality, or
   - The test generator - see tools on the project thread peNR48-2u-p2

**Testing instructions:**

1. Go to **WooCommerce → Settings → Payments → Woo Payments**
2. Disable the Stripe Billing setting.
4. You should see a warning notice pop up explaining that Subscriptions will be migrated.

Screenshot 2023-09-06 at 2 52 26 pm

5. Save settings.
6. A new notice should appear indicating that the migration is in-progress. 

Screenshot 2023-09-06 at 2 54 17 pm

7. Wait for the migration to complete.
    - if you refresh the settings page while the migration is in progress you'll continue to see that notice. 
    - You can track the migration on the **Tools → Scheduled Actions** screen. Search for `wcpay_schedule_subscription_migrations` and `wcpay_migrate_subscription` actions. 
9. Refresh the settings page and it should so the migration complete notice. 

Screenshot 2023-09-06 at 3 00 10 pm

</p>
</details>

<details><summary><strong>Choose migrate option</strong></summary>
<p>

**Setup:**

1. Woo Subscriptions inactive.
2. WCPay Subscriptions feature enabled. 
4. Delete the `_wcpay_feature_stripe_billing` option we want to simulate a merchant that has upgraded from WCPay Subscriptions to Woo Subscriptions for the first time.

**Testing instructions:**

1. Purchase a subscription (this will be a Stripe Billing subscription).
1. Enable the Woo Subscriptions plugin.
3. Go to **WooCommerce → Settings → Payments → Woo Payments**
4. You should see the following in the Advanced section.

Screenshot 2023-09-06 at 3 07 00 pm

5. Click the button to begin migration. 
6. You should see the in-progress notice.

Screenshot 2023-09-06 at 3 08 14 pm

7. Wait for the migration to complete. Refresh. 
9. You will see a completed notice. 

Screenshot 2023-09-06 at 3 11 34 pm

</p>
</details>

There's a lot of depth to these settings and the notices that are shown. They aren't all described above. 

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7105` or branch name `issue/6508` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=issue/6508) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: f06a68656c37616ebd4d0afad8c35b964052bf2d
- Build time: 2023-09-11 05:53:59 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/data/settings/selectors.js [hunk 1]
```diff
@@ -233,3 +233,19 @@ export const getAdvancedFraudProtectionSettings = ( state ) => {
 export const getShowWooPayIncompatibilityNotice = ( state ) => {
 	return getSettings( state ).show_woopay_incompatibility_notice || false;
 };
+
+export const getIsStripeBillingEnabled = ( state ) => {
+	return getSettings( state ).is_stripe_billing_enabled || false;
+};
+
+export const getIsStripeBillingMigrationInProcess = ( state ) => {
```
Comment (mattallan) [inline client/data/settings/selectors.js]: Should this be `getIsStripeBillingMigrationInProgress`?

```suggestion
export const getIsStripeBillingMigrationInProgress = ( state ) => {
```

File: client/settings/advanced-settings/stripe-billing-notices/migration-progress-notice.tsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
+/**
+ * External dependencies
+ */
+import React, { useState, useContext, useEffect } from 'react';
+import InlineNotice from 'wcpay/components/inline-notice';
+import { _n, sprintf } from '@wordpress/i18n';
+
+import Loadable from 'wcpay/components/loadable';
```
Comment (mattallan) [inline client/settings/advanced-settings/stripe-billing-notices/migration-progress-notice.tsx]: `Loadable` isn't used

File: client/settings/advanced-settings/stripe-billing-notices/migration-progress-notice.tsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	}, [ context.hasResolvedMigrateRequest ] );
+
+	// Set the notice to be eligible if Stripe Billing is saved as disabled. When disabling Stripe Billing, the migration will automatically start.
+	useEffect( () => {
+		if ( context.hasSavedSettings ) {
+			setIsEligible( ! context.savedIsStripeBillingEnabled );
+		}
+	}, [ context.hasSavedSettings, context.savedIsStripeBillingEnabled ] );
+
+	// Don't show the notice it's not eligible.
```
Comment (mattallan) [inline client/settings/advanced-settings/stripe-billing-notices/migration-progress-notice.tsx]:
```suggestion
	// Don't show the notice if it's not eligible.
```

**Comment by james-allan** [comment]
@diegocurbelo this is the PR I mentioned in our 1:1 last week. Do you mind giving the JS a review just to make sure I'm not breaking any obvious JS rules. 😄 

There's a lot going on in this PR and @mattallan is going to review the whole PR too.

**Review by mattallan — CHANGES_REQUESTED** [review]
I haven't done any testing of the settings yet but while reviewing the code and discussing [these line changes](https://github.com/Automattic/woocommerce-payments/pull/7105/files#diff-61d838bb73eba5daad8c16c1a6b535141ec17a004f0cf7e577b522c20f98d2b3L140-L142) specifically, we have came to the agreement that this area needs to be updated to cater for stores that have:
1. disabled the Stripe Billing feature and,
2. is in the middle of migrating over to tokenized subscriptions

For these cases, we would still want the Stripe Billing service classes to load so that merchants can continue to manage their Stripe Billing subscriptions before it's migrated, but also we continue to process any payments while they are in a pre-migrated state.

For this, we discussed two necessary changes to this:

1. Fixed in https://github.com/Automattic/woocommerce-payments/pull/7105/commits/5a89a37560e41c7e2a2d9b98dd849de79a8d24d0 ~~Inside the `WC_Payments_Subscriptions_Service::__construct()` function, bring back the [`if ( ! $this->is_subscriptions_plugin_active() )`](https://github.com/Automattic/woocommerce-payments/blob/4f75334f1cd97b77969b54db83a9a5fa62ada0d0/includes/subscriptions/class-wc-payments-subscription-service.php#L140) condition, but change it to check for `if ( WC_Payments_Features::is_wcpay_subscriptions_enabled() || WC_Payments_Features::is_stripe_billing_enabled() )`~~ 
2. Update the `should_load_wcpay_subscriptions()` so that we still load Stripe Billing/WCPay Subscription service classes if we are still migrating subscriptions, i.e. check if the store has existing subscriptions with `_wcpay_subscription_id` metadata. As well as remove the migration feature flag.

**Comment by mattallan** [comment]
Hey @james-allan I have a question: There are a handful of `! is_subscriptions_plugin_active()` checks found within the `includes/subscriptions` directory, should these be updated to

```
WC_Payments_Features::is_wcpay_subscriptions_enabled() || WC_Payments_Features::is_stripe_billing_enabled()
```

Some instances:
- [Product service construct](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/subscriptions/class-wc-payments-product-service.php#L96C3-L101C4)
- [limit_subscription_variation_intervals](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/subscriptions/class-wc-payments-product-service.php#L491)
- [limit_subscription_product_intervals](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/subscriptions/class-wc-payments-product-service.php#L449C18-L449C54)

I'm not sure if all of them can be replaced with checking for either wcpay or stripe billing settings, like some instances in the onboarding class.

**Comment by james-allan** [comment]
> Hey @james-allan I have a question: There are a handful of ! is_subscriptions_plugin_active() checks found within the includes/subscriptions directory, should these be updated to

Good spotting! 

After discussing this together I think we've landed on the right approach here now. We've both reviewed all the uses of `is_subscriptions_plugin_active()` and made a determination in each case if we need to update it. There  were some that should be changed to reflect whether Stripe Billing is in use or not. Some of the cases where we use the check `is_subscriptions_plugin_active()` should stay.

**Comment by mattallan** [comment]
### Testing

#### WCPay Subscriptions (legacy) + Woo Subscriptions deactivated

- [x] Confirm correct WooPayments setting is shown
- [x] Confirm disabling the setting removes Subscriptions features bundled in WooPayments
- [x] Purchase a subscription product, confirm Stripe Billing subscription
- [x] Process renewal with AS, confirm no payment is processed
- [x] Process renewal with billing clocks, confirm new charge and new order
- [x] Pay for failed order
- [x] Change payment method
- [x] Suspend and cancel subscription
  

#### Woo Subscriptions activated with existing WCPay Subscriptions

- [x] Confirm new Stripe Billing setting is deactivated by default with a begin migration option
- [x] From having WCPay Subscriptions to activating Woo Subscriptions, purchase a subscription and confirm it's tokenized.
- [x] Enable Stripe Billing and confirm existing tokenized subscription is still tokenized
- [x] Purchase a subscription product and confirm Stripe billing subscription is created
- [x] Process a scheduled action and confirm no payment is taken
- [x] Process renewal, change payment method, suspend and cancel subscription for the tokenized subscription
- [x] Process renewal, change payment method, suspend and cancel subscription for the Stripe Billing subscription
- [x] Pay for failed Stripe Billing renewal
  

#### Regular subscriptions /w Woo Subscriptions activated

- [x] Purchasing a subscription and confirm no Stripe Billing subscription is created
- [x] Processing tokenized renewal
    

#### Mid-migration Testing

- [x] While migrating, purchase a subscription and make sure it's not created as a Stripe Billing
- [x] With Stripe Billing disabled and subscriptions scheduled for migration, confirm the Stripe billing subscriptions continue to renew
  

#### Settings Notice

- [x] No notice when store is using Stripe Billing
- [x] "Begin Migration" notice shown on stores that have the setting disabled at page load but have existing WCPay Subscriptions
- [x] "Migration starting automatically" notice shown to stores that have existing WCPay Subscriptions and deactivate the feature
- [x] "Migration starting" notice shown when there's a migration in progress
- [x] "Migrated subscriptions" shows

**Review by mattallan — APPROVED** [review]
Awesome work on these changes @james-allan!
I've put this PR through a wide range of tests ([see this comment](https://github.com/Automattic/woocommerce-payments/pull/7105#issuecomment-1711147327)) to test the migration notices on the settings page but also to test stores in different states of the setting enabled and disabled and with existing WCPay Subscriptions, Stripe Billing subscriptions and tokenized 🥴

Let's ship it!

**Pull request merged**